src/app/clusters/electrical-power-measurement-server/electrical-power-measurement-server.cpp

```diff
@@ -0,0 +1,288 @@
+                            ChipLogError(Zcl, "Electrical Power Measurement: can not get RMSPower, feature is not supported"));
+        ReturnErrorOnFailure(aEncoder.Encode(mDelegate.GetRMSPower()));
+        break;
+    case Frequency::Id:
+        if (!SupportsOptAttr(OptionalAttributes::kOptionalAttributeFrequency))
+        {
+            return CHIP_IM_GLOBAL_STATUS(UnsupportedAttribute);
+        }
+        VerifyOrReturnError(HasFeature(ElectricalPowerMeasurement::Feature::kAlternatingCurrent),
+                            CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE,
+                            ChipLogError(Zcl, "Electrical Power Measurement: can not get Frequency, feature is not supported"));
+        ReturnErrorOnFailure(aEncoder.Encode(mDelegate.GetFrequency()));
+        break;
+    case HarmonicCurrents::Id:
+        return ReadHarmonicCurrents(aEncoder);
+    case HarmonicPhases::Id:
+        return ReadHarmonicPhases(aEncoder);
+    case PowerFactor::Id:
+        if (!SupportsOptAttr(OptionalAttributes::kOptionalAttributePowerFactor))
+        {
+            return CHIP_IM_GLOBAL_STATUS(UnsupportedAttribute);
+        }
+        VerifyOrReturnError(HasFeature(ElectricalPowerMeasurement::Feature::kAlternatingCurrent),
+                            CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE,
+                            ChipLogError(Zcl, "Electrical Power Measurement: can not get PowerFactor, feature is not supported"));
+        ReturnErrorOnFailure(aEncoder.Encode(mDelegate.GetPowerFactor()));
+        break;
+    case NeutralCurrent::Id:
+        if (!SupportsOptAttr(OptionalAttributes::kOptionalAttributeNeutralCurrent))
+        {
+            return CHIP_IM_GLOBAL_STATUS(UnsupportedAttribute);
+        }
+        VerifyOrReturnError(
+            HasFeature(ElectricalPowerMeasurement::Feature::kPolyphasePower), CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE,
+            ChipLogError(Zcl, "Electrical Power Measurement: can not get NeutralCurrent, feature is not supported"));
+        ReturnErrorOnFailure(aEncoder.Encode(mDelegate.GetNeutralCurrent()));
+        break;
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR Instance::ReadAccuracy(AttributeValueEncoder & aEncoder)
+{
+    CHIP_ERROR err  = CHIP_NO_ERROR;
+    auto accuracies = mDelegate.IterateAccuracy();
+    err             = aEncoder.EncodeList([&accuracies](const auto & encoder) -> CHIP_ERROR {
+        Structs::MeasurementAccuracyStruct::Type accuracy;
+        while (accuracies->Next(accuracy))
+        {
+            encoder.Encode(accuracy);
+        }
+
+        return CHIP_NO_ERROR;
+    });
+    accuracies->Release();
+    return err;
+}
+
+CHIP_ERROR Instance::ReadRanges(AttributeValueEncoder & aEncoder)
+{
+    if (!SupportsOptAttr(OptionalAttributes::kOptionalAttributeRanges))
+    {
+        return CHIP_IM_GLOBAL_STATUS(UnsupportedAttribute);
+    }
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    auto ranges    = mDelegate.IterateRanges();
```

Comment:
Again, can't this return null if the allocation fails?  We should return error, not crash, in that case.

Reply:
Changed to return error.